protobuf/src/main/scala/magnolify/protobuf/ProtobufType.scala

```diff
@@ -107,21 +122,23 @@ object ProtobufField {
       }
     }
 
-    override def to(v: T, bu: Message.Builder): Message =
-      caseClass.parameters
-        .foldLeft(bu.getDefaultInstanceForType.newBuilderForType()) { (b, p) =>
-          val field = b.getDescriptorForType.findFieldByName(p.label)
+    override def to(v: T, bu: Message.Builder): Message = {
+      val fields = getFields(bu.getDescriptorForType)
 
+      caseClass.parameters
+        .foldLeft(bu) { (b, p) =>
+          val field = fields(p.index)
           val value = if (field.getType == FieldDescriptor.Type.MESSAGE) {
             // nested records
             p.typeclass.to(p.dereference(v), b.newBuilderForField(field))
           } else {
             // non-nested
-            p.typeclass.to(p.dereference(v), b)
+            p.typeclass.to(p.dereference(v), null)
```

Comment:
Unrelated to the optimization, but builder is unused for primitives, setting it `null` to fail early.